src/web/services/datetime.service.ts

```diff
@@ -63,6 +63,32 @@ export class DateTimeService {
         return inst;
     }
 
+    getTimeFormatBeforeXMinutes(time: TimeFormat, x: number): TimeFormat {
+        let nhour = time.hour - Math.floor(x / 60);
+        let nmin = 0;
+        const nx = x % 60;
+        if (time.minute >= nx) {
+            nmin = time.minute - nx;
+        } else {
+            nhour -= 1;
+            nmin += 60 - x;
+        }
+        return { hour: nhour, minute: nmin };
+    }
```

Comment:
Let's not do it this way, working with time and date without a library is very error prone.

For example, `nhour = time.hour - Math.floor(x / 60)` can be negative here if time.hour is 0. `nhour -= 1; nmin += 60 - x;` doesn't look correct either, seems like this should be `60 - nx` instead.

The better way to do this would be to not work with TimeFormat and DateFormat directly. Could convert it to a js date object and add/subtract minutes from that instead.

Reply:
Addressed feedback. Let me know if there are any more required changes.